src/main/kotlin/org/eclipse/uprotocol/rpc/RpcMapper.kt

```diff
@@ -30,129 +30,72 @@ import com.google.protobuf.Message
-                exception = RuntimeException(
-                    String.format(
-                        "Unknown payload type [%s]. Expected [%s]",
-                        any.typeUrl, expectedClazz.getName()
-                    )
-                )
-                return@handle RpcResult.failure(exception.message!!, exception)
-            }
+/**
+ * Get the UStatus from Result&lt;Message&gt;.
+ * @return Returns UStatus if Result contains UStatus or Exception, returns Null if Result contains other payload.
+ * @param <T> The declared expected return type of the RPC method.
+</T> */
+fun <T : Message> Result<T>.getUStatusOrNull(): UStatus? {
+    return this.fold({ payload ->
+        if (payload is UStatus) {
+            payload
+        } else {
+            null
         }
-
-        @Suppress("UNCHECKED_CAST")
-        private fun <T : Message> calculateStatusResult(payload: Any): RpcResult<T> {
-            val status: UStatus = unpackPayload(payload, UStatus::class.java)
-            return if (status.code == UCode.OK) RpcResult.success(status as T) else RpcResult.failure(status)
+    }, { exception ->
+        uStatus {
+            code = UCode.UNKNOWN
+            message = exception.message ?: ""
         }
+    })
+}
 
-        /**
-         * Unpack a payload of type [Any] into an object of type T, which is what was packing into the [Any] object.
-         * @param payload an [Any] message containing a type of expectedClazz.
-         * @param expectedClazz The class name of the object packed into the [Any]
-         * @return Returns an object of type T and of the class name specified, that was packed into the [Any] object.
-         * @param <T> The message type of the object packed into the [Any].
-        </T> */
-        fun <T : Message> unpackPayload(payload: Any, expectedClazz: Class<T>): T {
-            return try {
-                payload.unpack(expectedClazz)
-            } catch (e: InvalidProtocolBufferException) {
-                throw RuntimeException(String.format("%s [%s]", e.message, UStatus::class.java.getName()), e)
-            }
-        }
+/**
+ * Get the UStatus from Result&lt;UPayload&gt;.
+ * @return Returns UStatus if Result contains UStatus or Exception.
+ */
+fun Result<UPayload>.mapUStatus(): UStatus {
+    return this.mapResponse(UStatus::class.java).getUStatusOrNull()?: uStatus {
+        code = UCode.UNKNOWN
+        message = "unexpected error"
+    }
+}
+
+/**
+ * Unpack a payload of type [Any] into an object of type T, which is what was packing into the [Any] object.
+ * @param payload an [Any] message containing a type of expectedClazz.
+ * @param expectedClazz The class name of the object packed into the [Any]
+ * @return Returns an object of type T and of the class name specified, that was packed into the [Any] object.
+ * @param <T> The message type of the object packed into the [Any].
+</T> */
+private fun <T : Message> unpackPayload(payload: Any, expectedClazz: Class<T>): T {
```

Comment:
Similarly, should we consider this `@PublishedApi internal inline fun <reified T: Message> unpackPayload(payload: Any): T` instead?